RxCocoa/Common/DelegateProxyFactory.swift

```diff
@@ -18,49 +18,71 @@ For example, in RxScrollViewDelegateProxy
 
 
     class RxScrollViewDelegateProxy: DelegateProxy {
-        static var factory = DelegateProxyFactory { (parentObject: UIScrollView) in
-            RxScrollViewDelegateProxy(parentObject: parentObject)
+        static var factory: DelegateProxyFactory {
+            return DelegateProxyFactory.sharedFactory(for: RxScrollViewDelegateProxy<UIScrollView>.self)
         }
     ...
 
 
-If need to extend them, chain `extended` after DelegateProxyFactory.init
+If need to extend them, call `DelegateProxySubclass.prepareForFactory()` in extend closure.
 
     class RxScrollViewDelegateProxy: DelegateProxy {
-        static var factory = DelegateProxyFactory { (parentObject: UIScrollView) in
-                RxScrollViewDelegateProxy(parentObject: parentObject)
-            }
-            .extended { (parentObject: UITableView) in
-                RxTableViewDelegateProxy(parentObject: parentObject)
+        static var factory: DelegateProxyFactory {
+            return DelegateProxyFactory.sharedFactory(for: RxScrollViewDelegateProxy<UIScrollView>.self) {
+                RxTableViewDelegateProxy<UITableView>.prepareForFactory()
             }
+        }
     ...
  
- 
+
  */
 public class DelegateProxyFactory {
+    private static var _sharedFactories: [ObjectIdentifier: DelegateProxyFactory] = [:]
+
+    /**
+     Shared instance of DelegateProxyFactory, if isn't exist shared instance, make DelegateProxyFactory instance for proxy type and extends.
+     DlegateProxyFactory have a shared instance per Delegate type.
+     - parameter proxyType: DelegateProxy type. Should use concrete DelegateProxy type, not generic.
+     - parameter extends: Extend DelegateProxyFactory if needs. See 'DelegateProxyType'.
+     - returns: DelegateProxyFactory shared instance.
+     */
+    public static func sharedFactory<DelegateProxy: DelegateProxyBase & DelegateProxyType>(for proxyType: DelegateProxy.Type, extends: (() -> Void)? = nil) -> DelegateProxyFactory {
```

Comment:
I think we should rename this. The name, intended usage and implementation are confusing to me.